src/core/mempool.rs

```diff
@@ -750,6 +756,113 @@ impl MemPoolTxInfo {
     }
 }
 
+/// Used to locally cache nonces to avoid repeatedly looking them up in the nonce.
+struct NonceCache {
+    cache: HashMap<StacksAddress, u64>,
+    size: usize,
+}
+
+impl NonceCache {
+    const MAX_SIZE: usize = 1024 * 1024;
+
+    fn new() -> Self {
+        Self {
+            cache: HashMap::new(),
+            size: 0,
+        }
+    }
+
+    fn get<C>(&mut self, address: &StacksAddress, clarity_tx: &mut C) -> u64
+    where
+        C: ClarityConnection,
+    {
+        match self.cache.get(address) {
+            Some(nonce) => *nonce,
+            None => {
+                let nonce = StacksChainState::get_nonce(clarity_tx, &address.clone().into());
+                // Simple size cap to the cache -- once it's full, all nonces
+                // will be looked up every time. This is bad for performance
+                // but is unlikely to occur due to the typical number of
+                // transactions processed before filling a block.
```

Comment:
How often is this cache cleared?  Is it once per block?

Also, it is knowable how many addresses can be loaded per block -- we could, in theory, calculate the maximum number of transactions that could be mined in a block, and use that to derive a maximum number of addresses that block could touch.

Reply:
Given how this cache gets used, it might make sense to just use a simple LRU strategy for now.  This cache is meant to help minimize the number of times we have to read the MARF to load up a nonce, so eliminating the most-common cases would be a good first attempt.

One day, subsequent refinement might consider the number of nodes that must be visited in the MARF to load the nonce.  If the address's nonce was recently changed, then there are fewer tries to visit.  It would make sense then to cache a nonce with probability proportional to how long ago it was last changed.  It doesn't have to be in this PR, but I'm flagging it here for consideration.

Reply:
@gregorycoppola I believe that because this cache is key to mempool iteration performance, we should get it checked in with this PR.

Reply:
@obycode On cache miss, could you instead write the last-known nonce to the mempool DB so we don't hit the MARF more than once per candidate?

Reply:
What if we just cap the cache size, and not do replacement?

That would be the simplest and provide the same benefit we have seen up to this point.

Reply:
Yes, this cache is cleared once per block.

> we could, in theory, calculate the maximum number of transactions that could be mined in a block, and use that to derive a maximum number of addresses that block could touch.

I don't think we can come up with a maximum number this way, because the number of addresses that it could touch is not bound by the number of transactions in the block, but by the number of transactions that need to be considered for the block. That would include pending transactions, and the number of those is only capped by the number of possible addresses.

Reply:
> On cache miss, could you instead write the last-known nonce to the mempool DB so we don't hit the MARF more than once per candidate on a call to iterate_candidates()?

Yes, we can do this. I'll benchmark again to see what kind of effect this has on performance in the normal case.

Reply:
LRU or no, I think you can bound transactions at around 256k using:

```
pub const MEMPOOL_MAX_TRANSACTION_AGE: u64 = 256;
```

And a calculation that around `1000` transactions fill a block.

So, the most number of transactions you would "expect" to crawl through before hitting a full block would be `256*1000=256k`.

Reply:
i was assuming that, on average, 1/256 transactions would have the right nonce.. i think that in practice the "proportion ready to mine" would actually go up and down a bit.

Reply:
also... the height that's used for eviction is independent of chaining, `MemPoolDB::tx_submit`... it just goes by 256 blocks since the height the tx was submitted, afaiu.

either way, this is just a model.